Gorilla - concurrent write to websocket connection

Got this error while using go-twitch-pubsub to read moderation events from Twitch
```
[go-twitch-pubsub] Lost connection, will try to reconnect
panic: concurrent write to websocket connection
goroutine 86520 [running]:
github.com/gorilla/websocket.(*messageWriter).flushFrame(0xc00015da40, 0x1, {0x0?, 0x0?, 0x0?})
         /root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:617 +0x52b
 github.com/gorilla/websocket.(*messageWriter).Close(0xc000382fc0?)
         /root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:731 +0x45
 github.com/gorilla/websocket.(*Conn).beginMessage(0xc000f44000, 0xc000382fc0, 0x1)
        /root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:480 +0x42
 github.com/gorilla/websocket.(*Conn).NextWriter(0xc000f44000, 0x1)
        /root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:520 +0x45
 github.com/gorilla/websocket.(*Conn).WriteMessage(0xc000035f98?, 0xc000035f78?, {0xc000d4ebb0, 0xf, 0x10})
         /root/go/pkg/mod/github.com/gorilla/websocket@v1.5.0/conn.go:773 +0x152
 github.com/pajlada/go-twitch-pubsub.(*connection).startWriter(0xc00007f540)
        /root/go/pkg/mod/github.com/pajlada/go-twitch-pubsub@v0.1.0/connection.go:223 +0x47
 created by github.com/pajlada/go-twitch-pubsub.(*connection).connect
        /root/go/pkg/mod/github.com/pajlada/go-twitch-pubsub@v0.1.0/connection.go:193 +0xf6
```
A quick Google search produced this, might be helpful: 
https://github.com/gorilla/websocket/issues/119

As for my code, I basically use the example:
```
func (a *twitchModeration) runPubSubClient() {
	pubsubClient := pubsub.NewClient(pubsub.DefaultHost)
	pubsubClient.Listen(pubsub.ModerationActionTopic(a.moderatorId, a.broadcasterId), a.config.Token)
	pubsubClient.OnModerationAction(func(channelID string, event *pubsub.ModerationAction) {
		a.processPubSubEvent(event)
	})

	go pubsubClient.Start()
}
````

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -29,6 +29,11 @@ jobs:
         with:
           go-version: ${{ matrix.go }}
 
+      - name: Install system dependencies
+        if: startsWith(matrix.os, 'macos-')
+        run: |
+          brew install docker
+
       - uses: actions/checkout@v4
 
       - name: Get Go dependencies

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -33,6 +33,7 @@ jobs:
         if: startsWith(matrix.os, 'macos-')
         run: |
           brew install docker
+          colima start
 
       - uses: actions/checkout@v4
 

--- a/.github/workflows/build.yml
+++ b/.github/workflows/build.yml
@@ -7,6 +7,9 @@ on:
       - master
   pull_request:
 
+env:
+  TWITCH_PUBSUB_SERVER_IMAGE: ghcr.io/chatterino/twitch-pubsub-server-test:v1.0.6
+
 jobs:
   build:
     runs-on: ${{ matrix.os }}
@@ -37,7 +40,10 @@ jobs:
         working-directory: cmd/example
 
       - name: Test
-        run: go test -v ./...
+        run: |
+          docker pull ${{ env.TWITCH_PUBSUB_SERVER_IMAGE }}
+          docker run --network=host --detach ${{ env.TWITCH_PUBSUB_SERVER_IMAGE }}
+          go test -v ./...
 
       - name: Generate code coverage
         if: matrix.os == 'ubuntu-22.04' && matrix.go == '1.19'

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -2,6 +2,7 @@
 
 ## Unreleased
 
+- Fix: Fix reconnect race condition on ping failure. (#33)
 - Dev: Add moderation event tests. (#21)
 - Dev: Add bits event tests. (#22, #23)
 

--- a/client.go
+++ b/client.go
@@ -7,12 +7,12 @@ import (
 )
 
 const (
-	reconnectInterval      = 5 * time.Second
-	pingInterval           = 4 * time.Minute
-	pongDeadlineTime       = 9 * time.Second
-	writerBufferLength     = 100
-	readerBufferLength     = 100
-	messageBusBufferLength = 50
+	defaultReconnectInterval = 5 * time.Second
+	defaultPingInterval      = 4 * time.Minute
+	defaultPongDeadlineTime  = 9 * time.Second
+	writerBufferLength       = 100
+	readerBufferLength       = 100
+	messageBusBufferLength   = 50
 
 	// maximum number of connections to open
 	defaultConnectionLimit = 10

--- a/connection.go
+++ b/connection.go
@@ -11,6 +11,12 @@ import (
 	"github.com/gorilla/websocket"
 )
 
+var connectionDialer *websocket.Dialer
+
+func setConnectionDialer(d *websocket.Dialer) {
+	connectionDialer = d
+}
+
 type connection struct {
 	host string
 
@@ -35,6 +41,14 @@ type connection struct {
 	topics []*websocketTopic
 
 	nonceCounter uint64
+
+	// numConnects gets incremented at the start of each connect call
+	// this is only used for tests
+	numConnects atomic.Uint64
+
+	reconnectInterval time.Duration
+	pingInterval      time.Duration
+	pongDeadlineTime  time.Duration
 }
 
 func newConnection(host string, messageBus messageBusType) *connection {
@@ -48,6 +62,10 @@ func newConnection(host string, messageBus messageBusType) *connection {
 		readerStop: make(chan bool),
 
 		messageBus: messageBus,
+
+		reconnectInterval: defaultReconnectInterval,
+		pingInterval:      defaultPingInterval,
+		pongDeadlineTime:  defaultPongDeadlineTime,
 	}
 }
 
@@ -65,6 +83,9 @@ func (c *connection) startReader() {
 		for {
 			messageType, payloadBytes, err := c.wsConn.ReadMessage()
 			if err != nil {
+				if websocket.IsUnexpectedCloseError(err, websocket.CloseGoingAway, websocket.CloseAbnormalClosure) {
+					fmt.Println("[go-twitch-pubsub]: Unexpected close error:", err)
+				}
 				c.readerStop <- true
 				return
 			}
@@ -75,11 +96,30 @@ func (c *connection) startReader() {
 		}
 	}()
 
+	pingTime := time.Now()
+	pingTicker := time.NewTicker(c.pingInterval)
+	pongCheckTimer := time.NewTimer(c.pongDeadlineTime)
+	pongCheckTimer.Stop()
+
 	for {
 		select {
 		case payloadBytes := <-c.reader:
 			if err := c.parse(payloadBytes); err != nil {
 				fmt.Println("Error parsing received websocket message:", err)
+			}
+
+		case <-pingTicker.C:
+			if err := c.ping(); err != nil {
+				fmt.Println("[go-twitch-pubsub] Error sending ping:", err)
+				return
+			}
+			pingTime = time.Now()
+			pongCheckTimer.Reset(c.pongDeadlineTime)
+
+		case <-pongCheckTimer.C:
+			if !c.lastPongWithinLimits(pingTime) {
+				fmt.Println("[go-twitch-pubsub] Lost connection, will try to reconnect")
+				return
 			}
 
 		case <-c.readerStop:
@@ -98,7 +138,8 @@ func (c *connection) lastPongWithinLimits(pingTime time.Time) bool {
 	c.pongMutex.Lock()
 	defer c.pongMutex.Unlock()
 
-	return c.lastPong.Sub(pingTime) < pongDeadlineTime
+	lastPongDiff := c.lastPong.Sub(pingTime)
+	return lastPongDiff >= 0*time.Second && lastPongDiff < c.pongDeadlineTime
 }
 
 func (c *connection) writeMessage(msg interface{}) error {
@@ -117,28 +158,12 @@ func (c *connection) ping() error {
 		Type: "PING",
 	}
 
-	pingTime := time.Now()
 	err := c.writeMessage(msg)
 	if err != nil {
 		return err
 	}
 
-	time.AfterFunc(pongDeadlineTime, func() {
-		if !c.lastPongWithinLimits(pingTime) {
-			fmt.Println("[go-twitch-pubsub] Lost connection, will try to reconnect")
-			c.doReconnect = true
-			c.onDisconnect()
-		}
-	})
-
 	return nil
-}
-
-func (c *connection) startPing() {
-	time.AfterFunc(pingInterval, func() {
-		_ = c.ping()
-		c.startPing()
-	})
 }
 
 func (c *connection) setConnected(newConnectedState bool) {
@@ -179,8 +204,13 @@ func (c *connection) stopReader() {
 }
 
 func (c *connection) connect() error {
+	c.numConnects.Add(1)
 	var err error
-	c.wsConn, _, err = websocket.DefaultDialer.Dial(c.host, nil)
+	if connectionDialer == nil {
+		c.wsConn, _, err = websocket.DefaultDialer.Dial(c.host, nil)
+	} else {
+		c.wsConn, _, err = connectionDialer.Dial(c.host, nil)
+	}
 	if err != nil {
 		c.doReconnect = true
 		c.onDisconnect()
@@ -192,13 +222,7 @@ func (c *connection) connect() error {
 	go c.startReader()
 	go c.startWriter()
 
-	c.onConnected()
-
 	return nil
-}
-
-func (c *connection) onConnected() {
-	go c.startPing()
 }
 
 func (c *connection) onDisconnect() {
@@ -207,7 +231,7 @@ func (c *connection) onDisconnect() {
 		return
 	}
 
-	time.AfterFunc(reconnectInterval, func() {
+	time.AfterFunc(c.reconnectInterval, func() {
 		c.tryReconnect()
 	})
 }

--- a/connection_test.go
+++ b/connection_test.go
@@ -1,0 +1,33 @@
+package twitchpubsub
+
+import (
+	"crypto/tls"
+	"testing"
+	"time"
+
+	qt "github.com/frankban/quicktest"
+	"github.com/gorilla/websocket"
+)
+
+func TestConnectionNoPingResponse(t *testing.T) {
+	c := qt.New(t)
+
+	// Run https://github.com/Chatterino/twitch-pubsub-server-test
+	dialer := *websocket.DefaultDialer
+	dialer.TLSClientConfig = &tls.Config{InsecureSkipVerify: true}
+	setConnectionDialer(&dialer)
+	messageBus := make(chan sharedMessage, 10)
+	conn := newConnection("wss://127.0.0.1:9050/dont-respond-to-ping", messageBus)
+	conn.pingInterval = 2 * time.Second
+	conn.pongDeadlineTime = 1 * time.Second
+	conn.pongDeadlineTime = 1 * time.Second
+
+	c.Assert(conn, qt.IsNotNil)
+
+	err := conn.connect()
+	c.Assert(err, qt.IsNil)
+
+	time.Sleep(10 * time.Second)
+
+	c.Assert(conn.numConnects.Load(), qt.Equals, uint64(2))
+}

